staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go

```diff
@@ -114,6 +126,32 @@ func (t *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 		}()
 		t.handler.ServeHTTP(w, rCopy)
 	}()
+
+	// Normally, the handler should notice the cancellation of the request
+	// context and stop. We give it some additional time and only then take
+	// over.
+	timeoutCh := make(chan struct{})
+
+	// This signals when ServeHTTP has returned, *not* when it is time to
+	// return. I.e. this intentionally doesn't use r.Context()!  We could
+	// wait for the timeout, but then the goroutine keeps running longer
+	// than it needs to.
+	done, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
```

Comment:
i am not sure, we need this additional goroutine, my understanding is: we have two goroutines that can try to read the request body at the same time, and that can only happen when a request times out:
- goroutine B: the is the goroutine onto which the timeout handler executes the rest of the handler chain.
- goroutine A: the is the goroutine net/http has spun up to execute the original handler chain.

in normal execution, A waits until B finishes, when B finishes A resumes. only when a timeout occurs, we see both A and B running concurrently. Maybe I am missing something here?

I have a proof PR https://github.com/kubernetes/kubernetes/pull/117164 that I believe reproduces the issue and solves it without this additional goroutine, please review, and feel free to use it if you see fit.

Reply:
Delaying the timeout handling in `timeoutHandler. ServeHTTP` is just some additional measure to make races less likely to occur. It cannot prevent them entirely. As you are planning to remove the forced timeout handling here, it's probably better to not make this code more complicated.